File: abstract_classes/abstract_classes_java/AbstractBinarySearchTree.java
package abstract_classes.abstract_classes_java;

public abstract class AbstractBinarySearchTree<T extends Comparable<T>> {
    
    // Initializes the abstract root node, inserting and deleting nodes
    protected Node<T> root;
    protected abstract Node<T> insert(Node<T> node, T value);
    protected abstract Node<T> delete(Node<T> node, T value);
    protected abstract boolean contains(Node<T> node, T value);

    // Allows us to traverse nodes from lowest to highest in order
    // (bottom left to bottom right)
    protected abstract void inOrderTraversal(Node<T> node);

    // Allows us to traverse nodes from root flowing down to left leaves then to right
    protected abstract void preOrderTraversal(Node<T> node);

    // Allows us to traverse nodes from left tree to right tree, saving each until end
    protected abstract void postOrderTraversal(Node<T> node);

    // Finds the smallest node value in the tree, traversing whole BST
    protected abstract Node<T> findMin(Node<T> node);

    // Finds the largest node value in the tree, traverses whole BST
    protected abstract Node<T> findMax(Node<T> node);

    // For each node, give it internal data (doesn't matter what data type, abstract as T)
    // Give each node the ability to talk with nodes to the left and to the right of it
    // (Inheritance, descendents as a concept)
    protected static class Node<T> {
        T data;
        Node<T> left;
        Node<T> right;

        public Node(T data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }
}

File: abstract_classes/abstract_classes_c++/AbstractBinarySearchTree.cpp
namespace abstract_classes {
    template<typename T>
    class AbstractBinarySearchTree {
    protected:
        struct Node {
            T data;
            Node* left;
            Node* right;

            Node(T data) : data(data), left(nullptr), right(nullptr) {}
        };

        Node* root;
        virtual Node* insert(Node* node, T value) = 0;
        virtual Node* deleteNode(Node* node, T value) = 0;
        virtual bool contains(Node* node, T value) = 0;
        virtual void inOrderTraversal(Node* node) = 0;
        virtual void preOrderTraversal(Node* node) = 0;
        virtual void postOrderTraversal(Node* node) = 0;
        virtual Node* findMin(Node* node) = 0;
        virtual Node* findMax(Node* node) = 0;

    public:
        virtual ~AbstractBinarySearchTree() = default;

        void insert(T value) {
            root = insert(root, value);
        }

        void remove(T value) {
            root = deleteNode(root, value);
        }

        bool contains(T value) {
            return contains(root, value);
        }

        void inOrderTraversal() {
            inOrderTraversal(root);
        }

        void preOrderTraversal() {
            preOrderTraversal(root);
        }

        void postOrderTraversal() {
            postOrderTraversal(root);
        }

        Node* findMin() {
            return findMin(root);
        }

        Node* findMax() {
            return findMax(root);
        }
    };
}
